sdks/python/apache_beam/runners/worker/statecache.py

```diff
@@ -0,0 +1,105 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""A module for caching state reads/writes in Beam applications."""
+from __future__ import absolute_import
+
+import collections
+from threading import Lock
+
+
+class StateCache(object):
+  """ Cache for Beam state access, scoped by state key and cache_token.
+
+  For a given state_key, caches a (cache_token, value) tuple and allows to
+    a) read from the cache,
+           if the currently stored cache_token matches the provided
+    a) write to the cache,
+           storing the new value alongside with a cache token
+    c) append to the cache cache,
+           if the currently stored cache_token matches the provided
+
+  The operations on the cache are thread-safe for use by multiple workers.
+  """
+
+  def __init__(self, max_entries):
+    self._cache = self.LRUCache(max_entries, (None, None))
+    self._lock = Lock()
+
+  def get(self, state_key, cache_token):
+    assert cache_token and self.is_cache_enabled()
+    with self._lock:
+      token, value = self._cache.get(state_key)
+    return value if token == cache_token else None
+
+  def put(self, state_key, cache_token, value):
+    assert cache_token and self.is_cache_enabled()
+    with self._lock:
+      return self._cache.put(state_key, (cache_token, value))
+
+  def append(self, state_key, cache_token, elements):
+    assert cache_token and self.is_cache_enabled()
+    with self._lock:
+      token, value = self._cache.get(state_key)
+      if token != cache_token:
+        # Discard cached state if tokens do not match
+        value = []
+      for element in elements:
+        value.append(element)
+      self._cache.put(state_key, (cache_token, value))
+
+  def clear(self, state_key):
+    assert self.is_cache_enabled()
+    with self._lock:
+      self._cache.clear(state_key)
```

Comment:
Clearing an entry from the cache is different than setting the value as cleared (but still being able to look it up). We want the latter, right?

Reply:
Correct, this would invalidate the cache which results in an unnecessary retrieval of the state from the Runner. We should instead just clear the value.